providers/dns/cloudflare/cloudflare.go

```diff
@@ -42,13 +46,22 @@ func NewDefaultConfig() *Config {
 
 // DNSProvider is an implementation of the challenge.Provider interface
 type DNSProvider struct {
-	client *cloudflare.API
+	dns    *cloudflare.API // needs Zone/DNS/Edit privileges
+	zone   *cloudflare.API // needs Zone/Zone/Read privileges
 	config *Config
+
+	zones   map[string]string
+	zonesMu *sync.RWMutex
 }
 
 // NewDNSProvider returns a DNSProvider instance configured for Cloudflare.
-// Credentials must be passed in the environment variables:
-// CLOUDFLARE_EMAIL, CLOUDFLARE_API_KEY, CLOUDFLARE_API_TOKEN.
+// Credentials must be passed in as environment variables:
+//
+// Either provide CLOUDFLARE_EMAIL and CLOUDFLARE_API_KEY,
+// or a CLOUDFLARE_DNS_API_TOKEN.
+// For a more paranoid setup, provide CLOUDFLARE_DNS_API_TOKEN and CLOUDFLARE_ZONE_API_TOKEN.
+//
+// Please see https://go-acme.github.com/lego/dns/cloudflare/ for a more detailed description.
```

Comment:
Is it OK to link to the website, or should I repeat what I wrote there?

Reply:
I prefer to not link to the website in the go code.

Reply:
but you don't need to repeat the documentation.

Reply:
Okay, I'm trying to condense the information a bit more.

Reply:
Done.